internal/core/src/common/File.h

```diff
@@ -94,16 +115,73 @@ class File {
     }
 
  private:
-    explicit File(int fd, const std::string& filepath)
-        : fd_(fd), filepath_(filepath) {
-        fs_ = fdopen(fd_, "wb+");
-        AssertInfo(fs_ != nullptr,
-                   "failed to open file {}: {}",
-                   filepath,
-                   strerror(errno));
+    explicit File(int fd, FILE* fs, const std::string& filepath)
+        : fd_(fd), filepath_(filepath), fs_(fs) {
     }
     int fd_{-1};
     FILE* fs_;
     std::string filepath_;
 };
+
+class BufferedWriter {
+ public:
+    // Constructor: Initialize with the file pointer and the buffer size (default 4KB).
+    explicit BufferedWriter(File& file, size_t buffer_size = 4096)
+        : file_(file),
+          buffer_size_(buffer_size),
+          buffer_(new char[buffer_size]) {
+    }
+
+    ~BufferedWriter() {
+        // Ensure the buffer is flushed when the object is destroyed
+        flush();
+        delete[] buffer_;
+    }
+
+    // Write method to handle data larger than the buffer
+    void
+    Write(const void* data, size_t size) {
+        if (size > buffer_size_) {
+            flush();
+            ssize_t written_data_size = file_.FWrite(data, size);
+            if (written_data_size != size) {
+                THROW_FILE_WRITE_ERROR
+            }
+            return;
+        }
+
+        if (buffer_pos_ + size > buffer_size_) {
+            flush();
+        }
+
+        std::memcpy(buffer_ + buffer_pos_, data, size);
+        buffer_pos_ += size;
+    }
+
+    template <typename T, std::enable_if_t<std::is_integral_v<T>, int> = 0>
+    void
+    WriteInt(T value) {
```

Comment:
why limit to int? should work on any plain type?

Reply:
just compatible with old interface, actually any plain type could share same code